### ISSUE TITLE
Design deep learning network

### ISSUE THREAD

**Issue by Samantha-Zhan** [YilikaLoufoua/noise-suppression#4]
Design the deep learning network using Deep Learning Toolbox. There are many network architectures to choose from. The two most common ones are convolutional neural networks, or CNN, ([[5]](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Speech%20Background%20Noise%20Suppression%20with%20Deep%20Learning#choi), [[6]](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Speech%20Background%20Noise%20Suppression%20with%20Deep%20Learning#isik)), and recurrent neural networks, or RNN ([[2]](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Speech%20Background%20Noise%20Suppression%20with%20Deep%20Learning#valin), [[3]](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Speech%20Background%20Noise%20Suppression%20with%20Deep%20Learning#xiang), [[4]](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Speech%20Background%20Noise%20Suppression%20with%20Deep%20Learning#nils)). Your solution may require applying signal processing at the input or output of your network (for example, for signal pre-processing, feature extraction, or time-frequency transformation). In that case, you can use Audio Toolbox and [Signal Processing Toolbox™](https://www.mathworks.com/products/signal.html) functionalities

**Comment by fissoreg** [comment]
As far as I know, you have a candidate architecture implemented. Where can I find the code? @Samantha-Zhan

**Comment by Samantha-Zhan** [comment]
I am encountering an error when I am running our deep learning model (image below, mainly is the dimension does not match for the concat layer). 
![image](https://user-images.githubusercontent.com/40647818/157747657-88272058-f54d-416b-88ee-35c1add009dd.png)
So I tried debugging and found the cause of it (the second and third dimension of in1 is inverted after passing through several layers, 128 is our mini-batch size)
![image](https://user-images.githubusercontent.com/40647818/157747850-4b65e20c-8946-4b8a-8767-f95141f6f8a9.png)
Then I was trying to apply permute function to in1 dlarray inside my custom reshpe layer, right before concat layer (custom concat is only there for testing purposes)
![image](https://user-images.githubusercontent.com/40647818/157748491-3669efb4-f6f7-48ec-a273-5cdfeeba7198.png)
![image](https://user-images.githubusercontent.com/40647818/157751913-ee04980d-9084-4a27-8751-3f2276d18933.png)
But due to matlab can't allow permutation across different channel (error below) I am really stuck now and not sure what to do. (I did also tried converting it to logical array and then back to dlarray but encountered error with dlgradient; so I concluded it is quite risky messing with differentiation and it might break the model)
![image](https://user-images.githubusercontent.com/40647818/157748799-b45f09f7-7cb3-402a-a73c-ad802274c782.png)

**Comment by Samantha-Zhan** [comment]
Code link: https://github.com/YilikaLoufoua/noise-suppression/tree/custom-training-loop/model
Problem: output of a custom layer got changed immediately after the layer returns
Detail: 
In the last part of our model, we wrote a custom layer normalLayer that does a series of manipulation to the input array. By setting a breakpoint inside the end of the custom layer, we know the output is in fact 257 313 32. However, when we check the dimension again inside checkdimsLayer (a custom layer that does nothing to the sequence but just provides a space to put breakpoint), we notice the dimension got changed by matlab to 32 1 313, with the Batch dimension reduced from 257 to 1.
![image](https://user-images.githubusercontent.com/40647818/160410826-aaacb7bf-9bdf-4ec5-8703-e2bb2faf8ba9.png)
Suspicion:
Our thought was that it has something to do with our dimension labelling inside the norm layer. Since we used reshape here to reduce down to 3D, reshape removes dimension label, so we had to manually add back the dimensions. It might be the way matlab processes dlarray after each layer that somehow removes all elements in the batch dimension.
![image](https://user-images.githubusercontent.com/40647818/160411949-f9733499-f3e8-44ee-a69f-84f478a600ae.png)
What we tried:
We have tried variable permutations of the letters B, C, T, S but none produces the correct result and we couldn't really tell what is happing inside Matlab's interworkings. Thus, we are quite stuck at this step.

**Comment by Samantha-Zhan** [comment]
normLayer -> 
CustomLayerAutodiffStrategy.m forwardExampleInputs function -> 
CustomLayerStrategy.m convertOutputsToPlaceholderArrays function -> 
CustomLayerStrategy.m iGetInputBatchSizeAndSequenceLength function
![image](https://user-images.githubusercontent.com/40647818/160626746-084eeefe-ef62-49c4-8159-8b28f4dc8110.png)
![image](https://user-images.githubusercontent.com/40647818/160626887-2cecb720-98e1-4fce-93b9-375a9a2cb9be.png)
![image](https://user-images.githubusercontent.com/40647818/160627144-27c87687-2096-47a7-9ed2-c882f63043d2.png)
![image](https://user-images.githubusercontent.com/40647818/160627357-c2e29efd-77db-4782-b1bb-425a643050d5.png)

**Comment by Samantha-Zhan** [comment]
**[UPDATE]** Dimension issue of normLayer (sort of?) resolved using unfold/fold layer of Matlab [here](https://github.com/YilikaLoufoua/noise-suppression/tree/custom-training-loop/model)
![image](https://user-images.githubusercontent.com/40647818/161114633-ff31c135-07e7-462b-99c9-232e0b7c34a3.png)
**[CURRENT ISSUE TO SOLVE]** The input does not meet the regression layer input requirement
**[WHAT WE TRIED]** We modified the fully connected layer output size temporarily to 1, so that after simple reshaping in final layer it gives 257(S) × 311(C) × 1(B) × 1(T) instead of 257(S) × 311(C) × 1(B) × 2(T). We also checked the input dimension to the network and it is 257(S) × 311(C) × 1(B) × 1(T), which matches exactly. However, only the first error goes away, second error remains(not sure what exactly is "expected input size" for regression layer?).
![image](https://user-images.githubusercontent.com/40647818/161114931-14a91cec-b095-4ae9-b0e8-a3c271b5d865.png)

**Comment by Samantha-Zhan** [comment]
**[UPDATE]** Regression layer mismatch issue resolved by removing T dimension labelling [here](https://github.com/YilikaLoufoua/noise-suppression/tree/custom-training-loop/model)

**[MY FIX OF THE ISSUE]** By going line by line into the source code, I found the function that throws the error(see below). By analyzing the code, I noticed it wouldn't throw an error if the input&output size matches exactly OR no T dimension. So all I did was that I changed the T labelling in the finalLayer to S. But then, as you said, I also have to change the fc_2 output from 2 to 1 for it to run with no error.
![image](https://user-images.githubusercontent.com/40647818/161598776-5882b1b5-eef6-40c9-8d27-6cb4d0f3e941.png)
![image](https://user-images.githubusercontent.com/40647818/161598797-dc6d430a-91ac-47ca-b32c-07ae1802845d.png)

**[RESULT AND NEW ISSUE]** after spending hours resolving some other backpropagation error, the model is finally able to train! However, it is training at a very slow speed (900min 152 iteration). And the training result is not very good but still decreasing (unit for bottom loss is 10*4)
![image](https://user-images.githubusercontent.com/40647818/161603408-036ce5b7-5e46-4c4b-b6b8-90d874852441.png)

**[POSSIBLE SOLUTION]** Giancarlo suggested this morning to compare the training result with running the researcher's PyTorch code. This can clarify whether there is something wrong with our model or just expected behaviour.

**Comment by Samantha-Zhan** [comment]
![image](https://user-images.githubusercontent.com/40647818/161657321-c74feab5-8a77-4a9b-94c9-ac08032b1648.png)
![image](https://user-images.githubusercontent.com/40647818/161657216-1d8c3ef1-c675-437b-9397-a08b9400f5ae.png)
